### 异常处理

####一、异常体系结构

java.lang.Throwable

* java.lang.Error：一般不编写针对性的代码进行处理
* java.lang.Exception：可以进行异常的处理
  * 编译时异常（checked）
    * IOException
      * FileNotFoundException
    * ClassNotFoundException
  * 运行时异常（unchecked,RuntimeException）
    * NullPointerException
    * ArrayIndexOutOfBoundsException
    * ClassCastException
    * InputMismatchException
    
    

<hr/>

#### 二、异常的处理：抓抛模型

1. “抓”：异常处理方式：①try-catch-finally ②throws
2. “抛”：程序在执行过程中，一旦出现异常，就会在异常代码处生成一个对应异常类的对象，并将此对象抛出，抛出之后的代码不再执行。



关于异常对象的产生：①系统自动生成的异常对象

​									  ②手动地生成一个异常对象，并抛出（throw）

```java
class StudentTest{
    public static void main(String[] args){
        try{
            Student s = new Student();
            s.regist(-1000);
        }catch(Exception e){
            System.out.println(e.getmessage());
        }
        
    }
}



class Student{
    
    private int id;
    
    public void regist(int id) throws Exception{
        if(id > 0){
            this.id = id;
        }else{
            //手动抛出异常对象
            //throw new RuntimeException("输入数据非法！");
            throw new Exception("输入数据非法！");
        }
    }
}
```





***注：方法重写与异常处理：***

​								**子类重写的方法抛出的异常类型不大于父类被重写的方法抛出的异常类型**



<hr/>

#### 三、异常处理的类型一：try-catch-finally



#####try-catch的使用

```java
try{
    //可能出现问题的代码
}catch(异常类型1 变量名1){
    //处理异常的方式1
}catch(异常类型2 变量名2){
    //处理异常的方式2
}
····
finally{
    //一定会执行的代码
}
    
```

注：

1. finally是可选的
2. 一旦try中的异常对象匹配到某一个catch时，就进入catch中进行异常的处理，处理完成，就跳出当前try-catch结构（无finally）
3. catch中的异常类型如果没有子父类关系，声明顺序无所谓；若满足，子类一定要声明在父类上面
4. 常用异常对象处理的方式：①String getMessage() ② printStackTrace()
5. 在try结构中声明的变量，出了try结构，就不能被调用

体会1：

​	使用try-catch-finally处理编译时异常，使程序在编译时不再报错，但是运行时仍可能报错，相当于我们将一个编译时可能出现的异常，延迟到运行时出现。

体会2：

​	开发中，由于运行时异常比较常见，所以我们通常不针对运行时异常编写try-catch-finally了，针对编译时异常，我们一定要考虑异常的处理



##### finally的使用

1. finally是可选的
2. finally中声明的是一定会被执行的代码。即使catch中又出现异常了，try中有return语句，catch中也有return语句等情况
3. 数据库连接，输入输出流，网络编程Socket等资源，JVM是不能自动回收的，我们需要自己手动的进行资源的释放，此时资源的释放，就需要声明在finally当中。

<hr/>

#### 四、异常处理的类型二：throws + 异常类型

1. "throws + 异常类型"写在方法的声明处。指明此方法执行时，可能会抛出的异常类型。一旦方法体执行时，出现异常，仍会在异常代码处生成一个异常类的对象，此对象满足throws后异常类型时，就会被抛出。异常代码后续的代码，就不再执行。
2. 体会：
   1. try-catch-finally：真正将异常处理掉了
   2. throws只是将异常抛给了方法的调用者，并没有真正处理掉
3. 开发中如何选择使用?
   1. 如果父类中被重写的方法没有throws方式处理异常，则子类重写的方法也不能使用throws，意味着如果子类重写的方法中有异常，必须使用try-catch-finally
   2. 执行的方法A中，先后又调用了另外的几个方法，这几个方法是递进关系执行的，我们建议这几个方法使用throws进行处理，执行的方法A可以考虑吧使用try-catch-finally方式进行处理



<hr/>

#### 五、自定义异常类

如何自定义异常类？

1. 继承于现有的异常结构：RuntimeException、Exception
2. 提供全局常量：serialVersionUID
3. 提供重载的构造器











